posts/2024-05-07-ultimate-string-guide.md

```diff
@@ -0,0 +1,625 @@
+all Chars to 8 bits. **You should avoid this, unless you know what you are doing.** It is more likely that you are looking
+for decoding libraries, where you can specify which encoding to use, e.g. [bytestring-encoding](https://hackage.haskell.org/package/bytestring-encoding-0.1.2.0/docs/Data-ByteString-Encoding.html#v:decode).
+
+It also has to be noted that pinned memory can cause memory fragmentation for lots of small ByteStrings
+(this is also discussed in [Fixing 'FilePath' in Haskell](https://hasufell.github.io/posts/2022-06-29-fixing-haskell-filepaths.html)).
+An alternative type is `ShortByteString`, which will be discussed next.
+
+#### ByteString summary
+
+Invariants:
+
+- pinned memory
+
+Useful for:
+
+- large data
+- dealing with raw bytes (e.g. web servers)
+- dealing with C FFI
+- storing non-unicode encodings e.g. via a newtype wrapper
+- fast parsers, see the excellent blog post from Chris Done on [Fast Haskell: Competing with C at parsing XML](https://chrisdone.com/posts/fast-haskell-c-parsing-xml/)
+
+Not so useful for:
+
+- dealing with unicode or human readable text
+- dealing with lots of small byte sequences
+
+Lazy variants, again, are useful for streaming and incremental processing, as the strict variant requires the whole content to be in memory.
+
+### ShortByteString
+
+This type is from the bytestring package as well and lives under
+[Data.ByteString.Short](https://hackage.haskell.org/package/bytestring-0.12.1.0/docs/Data-ByteString-Short.html).
+
+It has the same API as `ByteString` since [0.11.3.0](https://hackage.haskell.org/package/bytestring-0.11.3.0/changelog),
+so can be used as a drop-in replacement. The main difference is that it is *unpinned memory*, so causes no
+heap fragmentation. It also has no lazy variant.
+
+The definition as of [0.12.1.0](https://hackage.haskell.org/package/bytestring-0.12.1.0) is:
+
+```hs
+newtype ShortByteString =
+  ShortByteString
+  { unShortByteString :: ByteArray
+  }
+```
+
+This makes it suitable for things like unix filepaths. But we will explore better filepath types later.
+
+The name is maybe a little bit misleading. It can very well be used for large data as well, if you
+don't mind its strictness (the whole content is always in memory). **However, this type does not allow slicing**,
+unlike `Text` and `ByteString`, and so a lot of operations cause `memcpy`.
+
+Interfacing with C FFI triggers memory copy as well, because we need pinned memory.
+
+#### ShortByteString summary
+
+Invariants:
+
+- unpinned memory
+
+Useful for:
+
+- lots of small to medium sized byte sequences
+- large data, if strictness is desired and efficient slicing not needed
+- dealing with C FFI (although it incurs `memcpy`)
+- storing non-unicode encodings e.g. via a newtype wrapper
+
+Not so useful for:
+
+- dealing with unicode or human readable text
+- fast parsers, because no lazy variant and no efficient slicing
```

Comment:
It would be helpful to underline availability of efficient slicing for other types.

Reply:
Into how much detail should I go there? ByteString slicing is more efficient than Text (e.g. `splitAt`), but overall they are more efficient than Lists.

Reply:
I would go for **Text**:

> As we can see here, this type allows efficient slicing to avoid unnecessary `memcpy` for many operations.
E.g. `init` and `tail` are *O(1)* time and space. `splitAt` is *O(1)*, but *O(n)* time, because UTF-8 complicates
the offset computation.

And for **ByteString**:

> This allows, similar to Text, slicing without copying memory (through pointer arithmetic and the length field).
Since we're not dealing with unicode, but just `Word8` boundaries, operations like `splitAt` are *O(1)* time and
space.

And then consider Text *quite* efficient and ByteString *very* efficient for slicing.

Reply:
You probably meant "is O(1) __space__, but O(n) time"